include/netlink/netlink.h

```diff
@@ -32,6 +32,49 @@
 extern "C" {
 #endif
 
+/* macro to set selector's addr and prefix length;
+ * uses token pasting operator ## to paste either 'd' or 's' */
+#define xfrm_sel_to_xfrmnl_sel_for_src_or_dst(dst, src, what) { \
+	struct nl_addr *_addr; \
+	if ((src)->family == AF_INET) { \
+		_addr = nl_addr_build((src)->family, \
+				&(src)->what ## addr.a4, \
+				sizeof((src)->what ## addr.a4)); \
+	} else { \
+		_addr = nl_addr_build((src)->family, \
+				&(src)->what ## addr.a6, \
+				sizeof((src)->what ## addr.a6)); \
+	} \
+	nl_addr_set_prefixlen(_addr, (src)->prefixlen_ ## what); \
+	xfrmnl_sel_set_ ## what ## addr((dst), _addr); \
+	/* xfrmnl_sel_set_*addr increases the refcount; without this addr would leak */ \
+	nl_addr_put(_addr); \
+	xfrmnl_sel_set_prefixlen_ ## what((dst), (src)->prefixlen_ ## what); \
```

Comment:
I am sorry, such macros don't seem preferable to me.

Optimally, macros try to be very-function-like, that is:
- wrap in `do { } while (0)`
- evaluate all macro arguments exactly once (assigning to a local variable).
- avoid building names like `xfrmnl_sel_set_prefixlen_ ## what`, this means the name can no longer be found with `git grep` or `cscope`.

Granted, a function will always do something non-function-like (e.g. treat arguments of different types generically). Otherwise it should just be a C function. But the macro should aim to be function-like, unless there is a good reason not to.

But in this case, the macro does not seem an improvement.

I also don't understand, does this fix an actual bug, or does it just convert the code to a macro? The commit message should be clearer. If this is just about shortening code by putting it into a macro, then I am not convinced about the benefits of the result.